cmd/otk-resolve-containers/main_test.go

```diff
@@ -0,0 +1,194 @@
+
+	// add 10 images, all in the same repository with the same content
+	// (rootLayer), but each with a different tag and comment
+	refs := make([]string, 1)
+	for idx := 0; idx < len(refs); idx++ {
+		checksum := repo.AddImage(
+			[]testregistry.Blob{testregistry.NewDataBlobFromBase64(rootLayer)},
+			[]string{"amd64", "ppc64le"},
+			fmt.Sprintf("image %d", idx),
+			time.Time{})
+
+		tag := fmt.Sprintf("tag-%d", idx)
+		repo.AddTag(checksum, tag)
+		refs[idx] = fmt.Sprintf("%s:%s", ref, tag)
+	}
+	return registry, refs
+}
+
+func TestResolver(t *testing.T) {
+	registry, refs := createTestRegistry()
+	defer registry.Close()
+
+	inpContainers := make([]blueprint.Container, len(refs))
+	for idx, ref := range refs {
+		inpContainers[idx] = blueprint.Container{
+			Source:       ref,
+			Name:         fmt.Sprintf("test/localhost/%s", ref), // add a prefix for the local name to override the source
+			TLSVerify:    common.ToPtr(false),
+			LocalStorage: false,
+		}
+	}
+
+	for _, containerArch := range []string{"amd64", "ppc64le"} {
+		t.Run(containerArch, func(t *testing.T) {
+
+			require := require.New(t)
+			assert := assert.New(t)
+
+			amd64input := map[string]interface{}{
+				"arch":       containerArch,
+				"containers": inpContainers,
+			}
+			inputReq, err := json.Marshal(map[string]map[string]interface{}{
+				"tree": amd64input,
+			})
+			require.NoError(err)
+
+			inpBuf := bytes.NewBuffer(inputReq)
+			outBuf := &bytes.Buffer{}
+
+			assert.NoError(resolver.Run(inpBuf, outBuf))
+
+			var output map[string]resolver.Output
+			require.NoError(json.Unmarshal(outBuf.Bytes(), &output))
+
+			outputContainers := output["tree"]
+
+			assert.Len(outputContainers, len(refs))
+
+			expectedOutput := make([]resolver.ContainerInfo, len(refs))
+			for idx, ref := range refs {
+				// resolve directly with the registry and convert to ContainerInfo to
+				// compare with output.
+				spec, err := registry.Resolve(ref, arch.FromString(containerArch))
+				assert.NoError(err)
+				expectedOutput[idx] = resolver.ContainerInfo{
+					Source:  spec.Source,
+					Digest:  spec.Digest,
+					ImageID: spec.ImageID,
+					// registry.Resolve() copies the ref to the local name but the
+					// resolver will add the user-defined local name instead
+					LocalName:    fmt.Sprintf("test/localhost/%s", ref),
+					ListDigest:   spec.ListDigest,
+					Arch:         spec.Arch.String(),
+					LocalStorage: spec.LocalStorage,
+					TLSVerify:    spec.TLSVerify,
+				}
+			}
+
+			// NOTE: the order of containers in the resolver's output is stable but is
+			// not the same as the order of the inputs.
+			assert.ElementsMatch(outputContainers, expectedOutput)
+		})
+	}
+}
+
+func TestResolverUnhappy(t *testing.T) {
+	registry, refs := createTestRegistry()
+	defer registry.Close()
+
+	type testCase struct {
+		source    string
+		arch      string
+		tlsverify bool
+		errSubstr string
+	}
+
+	testCases := map[string]testCase{
```

Comment:
I wonder if we could get away with testing slightly less cases here? It's fine to keep as is as it's already written but at this level we are mostly caring about that when we call `Run()` and it results in an error but the details of the various ways that `run` can fail. Which seems mostly `resolver.Finish()` so testing that is great but at the same time just checking one way (or even mocking) is probably enough.

Reply:
I felt it was more useful to make sure that the error message provides useful information for a command line tool, vs simply testing that "error exists".
Of course, this all comes straight from `resolver.Finish()` like you mentioned, so maybe here is not the place to test for it.  I do like that we're making sure that _the container resolver command line tool isn't "eating" parts of the underlying error messages_.

Reply:
Sorry, I probably expressed this not very well, it's great to have a test that ensures we do pass the error properly and not "eat" part of the underlying error message. I was mostly wondering if we need to test this for all the cases. But as I wrote, it's fine as is.

Reply:
Not really necessary to cover all cases but I think it's nice to have.